### PR TITLE
[Draft] Perf design

### PR DESCRIPTION
**Pull request by taoran6**
Renamed classes:
DeviceTestTask->TestRun
ThreadParams->TestRunThreadContext

### CODE REVIEW

**Comment by hydraxman** [comment]
If our goal is to get a sequence of performance stats and the expected result is like a trending chart (memory trending) or a time sequence data, so in the PerformanceTestSpec there could be a field call profile sampling rate. Could we support this scenario?

File: sdk/src/test/java/com/microsoft/hydralab/appium/PerformanceTestSample.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.Arrays;
+
+import static com.microsoft.hydralab.performance.PerformanceTestSpec.*;
+
+/**
+ * Sample of performance test. Will not check in
+ */
+public class PerformanceTestSample {
+    @Test
+    public void performanceTestCase() {
```
Comment (taoran6) [inline sdk/src/test/java/com/microsoft/hydralab/appium/PerformanceTestSample.java:17]: Here is a sample of invoking performance inspection in UI test case

File: sdk/src/main/java/com/microsoft/hydralab/appium/ThreadParam.java [hunk 2]
```diff
@@ -2,21 +2,26 @@
...
 package com.microsoft.hydralab.appium;
 
+import com.microsoft.hydralab.performance.PerformanceInspectionService;
+
 import java.util.Map;
 
 public class ThreadParam {
     private static InheritableThreadLocal<AppiumParam> appiumParam = new InheritableThreadLocal<>();
     private static InheritableThreadLocal<Map<String, String>> configMap = new InheritableThreadLocal<>();
+    private static InheritableThreadLocal<PerformanceInspectionService> performanceExecutor = new InheritableThreadLocal<>();
```
Comment (taoran6) [inline sdk/src/main/java/com/microsoft/hydralab/appium/ThreadParam.java:12]: We will use `ThreadLocal` to share data between ` PerformanceInspectionService` and `TestRunner`.

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/TestRunner.java [hunk 3]
```diff
@@ -41,12 +48,29 @@ public void runTestOnDevice(TestTask testTask, DeviceInfo deviceInfo, Logger log
...
         try {
-            run(deviceInfo, testTask, deviceTestTask);
+            run(deviceInfo, testTask, deviceTestTask, performanceInspectionService);
         } catch (Exception e) {
             deviceTestTask.getLogger().error(deviceInfo.getSerialNum() + ": " + e.getMessage(), e);
             saveErrorSummary(deviceTestTask, e);
         } finally {
+            List<PerformanceResult<?>> performanceResults = performanceInspectionService.parse();
             tearDown(deviceInfo, testTask, deviceTestTask);
         }
```
Comment (taoran6) [inline agent/src/main/java/com/microsoft/hydralab/agent/runner/TestRunner.java:75]: Key codes of setup `PerformanceInspectionService`, starting performance inspection, and result analysis in `TestRunner`

File: common/src/main/java/com/microsoft/hydralab/common/entity/center/TestTaskSpec.java [hunk 4]
```diff
@@ -43,4 +43,6 @@ public class TestTaskSpec {
     public String testScope;
     public String testSuiteClass;
     public Map<String, List<DeviceAction>> deviceActions;
+    public boolean enableMemoryTest;
+    public boolean enableBatteryTest;
```
Comment (taoran6) [inline common/src/main/java/com/microsoft/hydralab/common/entity/center/TestTaskSpec.java:47]: Remove boolean

File: sdk/src/main/java/com/microsoft/hydralab/performance/PerformanceInspectionService.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    public void addInspector(PerformanceInspector performanceInspector) {
+        inspectors.add(performanceInspector);
+    }
+
+    public void initialize(PerformanceTestSpec performanceTestSpec) {
+        notifyEach(inspectors, recorder -> recorder.initialize(performanceTestSpec, resultFolder));
+    }
+
+    public void startInspectPerformanceTimer(PerformanceTestSpec performanceTestSpec, long interval) {
```
[the remark this answers is not in the document]
Comment (taoran6) [reply sdk/src/main/java/com/microsoft/hydralab/performance/PerformanceInspectionService.java:40]: 1. interval -> samplingRate./intervalInSecond
2. Move to PerformanceTestSpec

File: sdk/src/main/java/com/microsoft/hydralab/performance/PerformanceInspectionService.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+public class PerformanceInspectionService {
+    List<PerformanceInspector> inspectors = new ArrayList<>();
+    List<ScheduledFuture<?>> inspectPerformanceTimerList = new ArrayList<>();
+    List<PerformanceInspectionResult> performanceInspectionResultList = new ArrayList<>();
+    File resultFolder;
+    static final ScheduledExecutorService timerExecutor = Executors.newScheduledThreadPool(20 /* corePoolSize */);
```
Comment (taoran6) [inline sdk/src/main/java/com/microsoft/hydralab/performance/PerformanceInspectionService.java:19]: Refer to ThreadPoolTool. Change core size

File: sdk/src/main/java/com/microsoft/hydralab/performance/PerformanceResult.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    String category;
+
+    /**
+     * memory: java_heap_pss	java_heap_rss	native_heap_pss	native_heap_rss	code_pss	code_rss	stack_pss	stack_rss
+     * graphics_pss	graphics_rss	private_other_pss	private_other_rss	system_pss	system_rss	unknown_pss	unknown_rss
+     * total_pss	total_rss	total_swap_pss
+     * <p>
+     * battery: CPU	screen	Wake_lock	other	App_usage	Total_usage
+     */
+    T performanceData;
```
Comment (taoran6) [inline sdk/src/main/java/com/microsoft/hydralab/performance/PerformanceResult.java:15]: performanceSummaryResultsData {
   List<> performanceInspectionResult
  average_java_heap_pss
  max_java_heap_pss

}

File: sdk/src/main/java/com/microsoft/hydralab/performance/PerformanceTestSpec.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    public File resultFolder;
+    public long interval;
+
+    public PerformanceTestSpec(String inspector, String appId, String deviceId, String name) {
+        this.inspector = inspector;
+        this.appId = appId;
+        this.deviceId = deviceId;
+        this.name = name;
+        inspectionKey = String.format("%s-%s-%s", appId, deviceId, inspector);
```
Comment (taoran6) [inline sdk/src/main/java/com/microsoft/hydralab/performance/PerformanceTestSpec.java:29]: Do we support simultaneous performance testing on different android devices? If the device id is "Android", we may not tell them apart.

File: common/src/main/java/com/microsoft/hydralab/performance/PerformanceInspectorManagementService.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.*;
+
+@Service
+public class PerformanceInspectorManagementService implements IPerformanceInspectionService {
+    static final ScheduledExecutorService timerExecutor = Executors.newScheduledThreadPool(5 /* corePoolSize */);
```
Comment (taoran6) [inline common/src/main/java/com/microsoft/hydralab/performance/PerformanceInspectorManagementService.java:13]: Can the type of inspectPerformanceTimerList also be `Map<String, ScheduledFuture<?>>`?

File: sdk/src/main/java/com/microsoft/hydralab/performance/InspectionStrategy.java [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package com.microsoft.hydralab.performance;
+
+import java.util.concurrent.TimeUnit;
+
+public class InspectionStrategy {
+    public int interval;
+    public TimeUnit intervalUnit;
+    public String when;
+    public int whenType;
```
Comment (taoran6) [inline sdk/src/main/java/com/microsoft/hydralab/performance/InspectionStrategy.java:9]: What are these two fields for?

File: agent/doc/UML/perf_design.puml [hunk 11]
```diff
@@ -0,0 +1,164 @@
+@startuml
+'https://plantuml.com/class-diagram'
+
+package "sdk" {
+
+    class PerformanceTestSpec {
```
Comment (hydraxman) [inline agent/doc/UML/perf_design.puml:6]:
```suggestion
    class PerformanceInspection {
```

File: agent/doc/UML/agent_device_management_design.puml [hunk 12]
```diff
@@ -27,4 +27,25 @@ ADBDriver -u-|> TestDriver
 AppiumDriver -u-|> TestDriver
 
 TestRunner -r-> TestDeviceManager
+@enduml
+
+@startuml test_device
+'https://plantuml.com/class-diagram
+interface ITestDevice {
+    List<ITestDevice> getAllTestDevices();
```
Comment (hydraxman) [inline agent/doc/UML/agent_device_management_design.puml:35]:
```suggestion
```

File: agent/src/main/java/com/microsoft/hydralab/agent/config/TestRunnerConfig.java [hunk 13]
```diff
@@ -33,39 +34,45 @@ public class TestRunnerConfig {
...
+        PerformanceTestManagementService performanceTestManagementService = new PerformanceTestManagementService();
+        performanceTestManagementService.initialize();
+        return performanceTestManagementService;
+    }
 
     @Bean
-    public EspressoRunner espressoRunner(DeviceManager deviceManager, TestTaskEngineService testTaskEngineService, ADBOperateUtil adbOperateUtil) {
-        return new EspressoRunner(deviceManager, testTaskEngineService, adbOperateUtil);
+    public EspressoRunner espressoRunner(DeviceManager deviceManager, TestTaskEngineService testTaskEngineService, ADBOperateUtil adbOperateUtil, PerformanceTestManagementService performanceTestManagementService) {
+        return new EspressoRunner(deviceManager, testTaskEngineService, adbOperateUtil, performanceTestManagementService);
```
Comment (hydraxman) [inline agent/src/main/java/com/microsoft/hydralab/agent/config/TestRunnerConfig.java:46]:
```suggestion
    public EspressoRunner espressoRunner(DeviceManager deviceManager, TestTaskEngineService testTaskEngineService, ADBOperateUtil adbOperateUtil, PerformanceTestManagementService performanceTestManagementService) {
        return new TestRunner(deviceManager, testTaskEngineService, adbOperateUtil, performanceTestManagementService);
```

**Pull request closed without merging**